graspologic/cluster/rclust.py

```diff
@@ -0,0 +1,377 @@
+            raise ValueError(msg)
+
+        if delta_criter is not None:
+            if delta_criter <= 0:
+                raise ValueError("delta_criter must be positive")
+
+        if likelihood_ratio is not None:
+            if likelihood_ratio <= 0 or likelihood_ratio >= 1:
+                raise ValueError("likelihood_ratio must be in (0,1)")
+
+        self.parent = parent
+        self.min_components = min_components
+        self.max_components = max_components
+        self.cluster_method = cluster_method
+        self.min_split = min_split
+        self.selection_criteria = selection_criteria
+        self.cluster_kws = cluster_kws
+        self.min_split = min_split
+        self.max_level = max_level
+        self.delta_criter = delta_criter
+        self.likelihood_ratio = likelihood_ratio
+
+    def fit(self, X):
+        """
+        Fits clustering models to the data as well as resulting clusters
+
+        Parameters
+        ----------
+        X : array_like, shape (n_samples, n_features)
+
+        Returns
+        -------
+        self : object
+            Returns an instance of self.
+        """
+        self.fit_predict(X)
+        return self
+
+    def fit_predict(self, X, level=None):
+        """
+        Fits clustering models to the data as well as resulting clusters
+        and using fitted models to predict a hierarchy of labels
+
+        Parameters
+        ----------
+        X : array_like, shape (n_samples, n_features)
+        level: int, optional (default=None)
+
+        Returns
+        -------
+        labels : array_label, shape (n_samples, n_levels)
+            if no level specified; otherwise, shape (n_samples,)
+        """
+        X = check_array(X, dtype=[np.float64, np.float32], ensure_min_samples=1)
+        if level is not None:
+            if not isinstance(level, int):
+                raise TypeError("level must be an int")
+            elif level < 1:
+                raise ValueError("level must be positive")
+
+        self.X = X
+        self.labels = np.zeros((len(X), 1), dtype=int)
+        self.indx = np.arange(len(X))
+
+        if self.max_components > X.shape[0]:
+            msg = "max_components must be >= n_samples, but max_components = "
+            msg += "{}, n_samples = {}".format(self.max_components, X.shape[0])
+            raise ValueError(msg)
+
+        while True:
+            current_node = self._get_next_node()
+            if current_node:
+                current_node._cluster_and_decide()
+            else:
+                break
+
+        labels = self._unique_labels(self.labels)
+        if level is not None:
+            if level <= labels.shape[1]:
+                labels = labels[:, level - 1]
+            else:
+                msg = "input exceeds max level = {}".format(labels.shape[1])
+                raise ValueError(msg)
+
+        self.predictions = labels
+
+        return labels
+
+    def _get_next_node(self):
```

Comment:
encourage you to use one of the tree traversal methods in anytree if possible. I bet there is a way to use those instead of this. maybe worth looking into at least?